File: scripts/memory_config.py
#!/usr/bin/env python3
"""解析记忆脚本使用的外部配置。"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path


CONFIG_PATH = Path.home() / ".config" / "memorymanager" / "config.json"
DEFAULT_STORAGE_ROOT = Path.home() / ".local" / "share" / "memorymanager"
STORAGE_PATH_KEYS = (
    "memory_storage_path",
    "memoryStorePath",
    "storage_path",
    "storagePath",
)
PROJECTS_KEYS = (
    "external_projects",
    "externalProjects",
    "projects",
)


@dataclass(frozen=True)
class MemoryLocation:
    """统一描述当前项目对应的记忆位置与项目标识。"""

    project_root: Path
    project_name: str
    search_root: Path
    memory_root: Path
    external_enabled: bool
    config_path: Path


def resolve_memory_location(project_root: Path) -> MemoryLocation:
    """根据项目根目录与用户配置决定实际记忆目录。"""

    resolved_root = project_root.resolve()
    project_name = sanitize_project_name(resolved_root.name)
    config = load_config(CONFIG_PATH)

    if config and project_in_external_list(config, resolved_root, project_name):
        storage_root = resolve_storage_root(config, CONFIG_PATH.parent)
        if storage_root is not None:
            return MemoryLocation(
                project_root=resolved_root,
                project_name=project_name,
                search_root=storage_root,
                memory_root=storage_root / ".memory",
                external_enabled=True,
                config_path=CONFIG_PATH,
            )

    return MemoryLocation(
        project_root=resolved_root,
        project_name=project_name,
        search_root=resolved_root,
        memory_root=resolved_root / ".memory",
        external_enabled=False,
        config_path=CONFIG_PATH,
    )


def sanitize_project_name(name: str) -> str:
    """保证项目名稳定可用，避免空名或非法片段污染检索维度。"""

    cleaned = name.strip().strip("./")
    return cleaned or "default-project"


def load_config(config_path: Path) -> dict[str, object] | None:
    """读取配置文件，缺失时自动创建默认配置以降低首次使用门槛。"""

    if not config_path.exists():
        default_config = build_default_config()
        if write_default_config(config_path, default_config):
            return default_config
        return None
    try:
        payload = json.loads(config_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    return payload if isinstance(payload, dict) else None


def build_default_config() -> dict[str, object]:
    """生成默认配置，提前给出统一的外挂记忆落盘位置。"""

    return {
        "memory_storage_path": str(DEFAULT_STORAGE_ROOT),
        "external_projects": [],
    }


def write_default_config(config_path: Path, config: dict[str, object]) -> bool:
    """首次运行自动落默认配置，避免用户必须手写样板文件。"""

    try:
        config_path.parent.mkdir(parents=True, exist_ok=True)
        config_path.write_text(
            json.dumps(config, ensure_ascii=False, indent=2) + "\n",
            encoding="utf-8",
        )
    except OSError:
        return False
    return True


def project_in_external_list(
    config: dict[str, object], project_root: Path, project_name: str
) -> bool:
    """兼容项目名、项目根路径和对象写法，尽量减少配置格式耦合。"""

    entries = extract_projects(config)
    project_root_text = str(project_root)
    for entry in entries:
        if isinstance(entry, str):
            normalized = entry.strip()
            if normalized in {project_name, project_root_text}:
                return True
            continue

        if not isinstance(entry, dict):
            continue

        candidate_name = first_text(entry, "name", "project", "project_name")
        if candidate_name and candidate_name == project_name:
            return True

        candidate_path = first_text(entry, "path", "root", "project_root")
        if candidate_path:
            candidate_root = Path(candidate_path).expanduser()
            if not candidate_root.is_absolute():
                candidate_root = (CONFIG_PATH.parent / candidate_root).resolve()
            else:
                candidate_root = candidate_root.resolve()
            if candidate_root == project_root:
                return True
    return False


def extract_projects(config: dict[str, object]) -> list[object]:
    """从多个兼容字段中提取外挂项目列表。"""

    for key in PROJECTS_KEYS:
        value = config.get(key)
        if isinstance(value, list):
            return value
    return []


def resolve_storage_root(config: dict[str, object], base_dir: Path) -> Path | None:
    """解析外挂记忆根目录，支持相对路径配置。"""

    for key in STORAGE_PATH_KEYS:
        value = config.get(key)
        if not isinstance(value, str) or not value.strip():
            continue
        storage_root = Path(value).expanduser()
        if not storage_root.is_absolute():
            storage_root = (base_dir / storage_root).resolve()
        else:
            storage_root = storage_root.resolve()
        return storage_root
    return None


def first_text(payload: dict[str, object], *keys: str) -> str:
    """从候选键中取第一个非空字符串。"""

    for key in keys:
        value = payload.get(key)
        if isinstance(value, str) and value.strip():
            return value.strip()
    return ""

File: memory-manager/scripts/memory_config.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// EmbeddingConfig 统一描述嵌入配置，避免不同子命令各自解析字段。
type EmbeddingConfig struct {
	BaseURL        string
	APIKey         string
	Model          string
	TimeoutSeconds float64
}

// MemoryLocation 统一描述当前项目对应的记忆目录与检索维度。
type MemoryLocation struct {
	ProjectRoot     string
	ProjectName     string
	SearchRoot      string
	MemoryRoot      string
	ExternalEnabled bool
	ConfigPath      string
	EmbeddingConfig *EmbeddingConfig
}

var (
	storagePathKeys      = []string{"memory_storage_path", "memoryStorePath", "storage_path", "storagePath"}
	embeddingSectionKeys = []string{"embedding", "embeddings"}
	baseURLKeys          = []string{"base_url", "baseUrl", "url", "endpoint"}
	apiKeyKeys           = []string{"api_key", "apiKey"}
	modelKeys            = []string{"model", "embedding_model", "embeddingModel"}
	timeoutKeys          = []string{"timeout_seconds", "timeoutSeconds"}
)

// resolveMemoryLocation 优先项目内目录，再尝试外挂目录，最后回退到本地目录。
func resolveMemoryLocation(projectRoot string) (MemoryLocation, error) {
	resolvedRoot, err := filepath.Abs(projectRoot)
	if err != nil {
		return MemoryLocation{}, err
	}
	projectName := sanitizeProjectName(filepath.Base(resolvedRoot))
	configPath, err := defaultConfigPath()
	if err != nil {
		return MemoryLocation{}, err
	}
	config, _ := loadConfig(configPath)
	defaultConfig, err := buildDefaultConfig()
	if err != nil {
		return MemoryLocation{}, err
	}
	if config == nil {
		config = defaultConfig
	}
	embeddingConfig := resolveEmbeddingConfig(config)
	localMemoryRoot := filepath.Join(resolvedRoot, ".memory")

	if isDir(localMemoryRoot) {
		return MemoryLocation{
			ProjectRoot:     resolvedRoot,
			ProjectName:     projectName,
			SearchRoot:      resolvedRoot,
			MemoryRoot:      localMemoryRoot,
			ExternalEnabled: false,
			ConfigPath:      configPath,
			EmbeddingConfig: embeddingConfig,
		}, nil
	}

	storageRoot := resolveStorageRoot(config, filepath.Dir(configPath))
	if storageRoot != "" {
		externalMemoryRoot := filepath.Join(storageRoot, ".memory")
		if ensureDirUsable(externalMemoryRoot) == nil {
			return MemoryLocation{
				ProjectRoot:     resolvedRoot,
				ProjectName:     projectName,
				SearchRoot:      storageRoot,
				MemoryRoot:      externalMemoryRoot,
				ExternalEnabled: true,
				ConfigPath:      configPath,
				EmbeddingConfig: embeddingConfig,
			}, nil
		}
	}

	return MemoryLocation{
		ProjectRoot:     resolvedRoot,
		ProjectName:     projectName,
		SearchRoot:      resolvedRoot,
		MemoryRoot:      localMemoryRoot,
		ExternalEnabled: false,
		ConfigPath:      configPath,
		EmbeddingConfig: embeddingConfig,
	}, nil
}

// defaultConfigPath 统一配置文件路径，减少路径散落导致的行为差异。
func defaultConfigPath() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, ".config", "memorymanager", "config.json"), nil
}

// buildDefaultConfig 提供默认外挂目录，兼容首次运行场景。
func buildDefaultConfig() (map[string]any, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil, err
	}
	return map[string]any{
		"memory_storage_path": filepath.Join(home, ".local", "share", "memorymanager"),
		"embedding": map[string]any{
			"base_url":        "",
			"api_key":         "",
			"model":           "",
			"timeout_seconds": 30,
		},
	}, nil
}

// loadConfig 读取配置文件，缺失时尽量创建默认配置，但不因失败中断主流程。
func loadConfig(configPath string) (map[string]any, error) {
	content, err := os.ReadFile(configPath)
	if err != nil {
		if !errors.Is(err, os.ErrNotExist) {
			return nil, err
		}
		defaultConfig, buildErr := buildDefaultConfig()
		if buildErr != nil {
			return nil, buildErr
		}
		if writeDefaultConfig(configPath, defaultConfig) == nil {
			return defaultConfig, nil
		}
		return nil, nil
	}
	var payload map[string]any
	if err := json.Unmarshal(content, &payload); err != nil {
		return nil, err
	}
	return payload, nil
}

// writeDefaultConfig 仅在首次运行时落默认配置，避免用户必须手写样板。
func writeDefaultConfig(configPath string, config map[string]any) error {
	if err := os.MkdirAll(filepath.Dir(configPath), 0o755); err != nil {
		return err
	}
	content, err := json.MarshalIndent(config, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(configPath, append(content, '\n'), 0o644)
}

// resolveStorageRoot 解析外挂目录并兼容相对路径配置。
func resolveStorageRoot(config map[string]any, baseDir string) string {
	for _, key := range storagePathKeys {
		value := strings.TrimSpace(asString(config[key]))
		if value == "" {
			continue
		}
		if strings.HasPrefix(value, "~") {
			home, err := os.UserHomeDir()
			if err == nil {
				value = filepath.Join(home, strings.TrimPrefix(value, "~/"))
			}
		}
		if filepath.IsAbs(value) {
			return filepath.Clean(value)
		}
		return filepath.Clean(filepath.Join(baseDir, value))
	}
	return ""
}

// resolveEmbeddingConfig 提取嵌入配置，字段不完整时视为关闭。
func resolveEmbeddingConfig(config map[string]any) *EmbeddingConfig {
	section := findEmbeddingSection(config)
	if section == nil {
		return nil
	}
	baseURL := strings.TrimRight(pickString(section, baseURLKeys), "/")
	apiKey := pickString(section, apiKeyKeys)
	model := pickString(section, modelKeys)
	if baseURL == "" || apiKey == "" || model == "" {
		return nil
	}
	timeout := pickFloat(section, timeoutKeys, 30)
	if timeout < 1 {
		timeout = 1
	}
	return &EmbeddingConfig{
		BaseURL:        baseURL,
		APIKey:         apiKey,
		Model:          model,
		TimeoutSeconds: timeout,
	}
}

// findEmbeddingSection 优先取嵌套配置，必要时兼容平铺字段。
func findEmbeddingSection(config map[string]any) map[string]any {
	for _, key := range embeddingSectionKeys {
		if payload, ok := config[key].(map[string]any); ok {
			return payload
		}
	}
	return config
}

// pickString 从多个候选键中取第一个非空字符串，减少重复解析代码。
func pickString(payload map[string]any, keys []string) string {
	for _, key := range keys {
		value := strings.TrimSpace(asString(payload[key]))
		if value != "" {
			return value
		}
	}
	return ""
}

// pickFloat 对数字和字符串都做兼容解析，避免配置格式差异导致失效。
func pickFloat(payload map[string]any, keys []string, defaultValue float64) float64 {
	for _, key := range keys {
		switch value := payload[key].(type) {
		case float64:
			return value
		case float32:
			return float64(value)
		case int:
			return float64(value)
		case int64:
			return float64(value)
		case json.Number:
			if v, err := value.Float64(); err == nil {
				return v
			}
		case string:
			if v, err := strconv.ParseFloat(strings.TrimSpace(value), 64); err == nil {
				return v
			}
		}
	}
	return defaultValue
}

// sanitizeProjectName 保证项目名稳定，避免空名污染共享库维度。
func sanitizeProjectName(name string) string {
	cleaned := strings.Trim(strings.TrimSpace(name), "./")
	if cleaned == "" {
		return "default-project"
	}
	return cleaned
}

// ensureDirUsable 尝试创建目录来判断目标位置是否可用，优先避免运行时权限错误。
func ensureDirUsable(path string) error {
	return os.MkdirAll(path, 0o755)
}

// isDir 用显式目录判断来保留项目内 `.memory` 的优先级语义。
func isDir(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

// asString 统一把动态值转成字符串，方便做宽松兼容。
func asString(value any) string {
	switch v := value.(type) {
	case string:
		return v
	case json.Number:
		return v.String()
	case nil:
		return ""
	default:
		return strings.TrimSpace(fmt.Sprint(v))
	}
}
